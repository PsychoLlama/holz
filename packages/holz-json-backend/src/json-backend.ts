import type { Writable } from 'node:stream';
import { EOL } from 'node:os';
import type { Log, LogProcessor } from '@holz/core';

/**
 * Prints structured logs to a writable stream in NDJSON form. Optimized for
 * log files.
 *
 * @example
 * new JsonBackend({
 *   stream: fs.createWriteStream('my-app.log', { flags: 'a' }),
 * })
 *
 * @see http://ndjson.org
 */
export default class JsonBackend implements LogProcessor {
  private stream: Writable;

  constructor(options: Config) {
    this.stream = options.stream;
  }

  processLog(log: Log) {
    // Follow the order of typical log statements. Be kind to the human
    // reader.
    const output = JSON.stringify({
      level: log.level,
      time: new Date().toISOString(),
      msg: log.message,
      ctx: Object.keys(log.context).length > 0 ? log.context : undefined,
    });

    // NOTE: If the stream applies backpressure, we will lose logs. I believe
    // this is the right tradeoff. We can't prevent the app from generating
    // more logs, and if we buffered it would risk running out of memory and
    // crashing the process.
    //
    // It is unlikely that a file or tty will apply backpressure in practice.
    this.stream.write(`${output}${EOL}`);
  }
}

interface Config {
  /** Where to print logs. */
  stream: Writable;
}
