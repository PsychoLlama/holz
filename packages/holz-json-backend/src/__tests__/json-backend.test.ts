import { Writable } from 'node:stream';
import { createLogger } from '@holz/core';
import { createJsonBackend } from '../json-backend';

const CURRENT_TIME = new Date('2020-06-15T12:00:00.000Z');

describe('JSON backend', () => {
  function createStream() {
    let output = '';
    const stream = new Writable({
      write(chunk, _encoding, callback) {
        output += String(chunk);
        callback();
      },
    });

    return {
      getOutput: () => output,
      stream,
    };
  }

  beforeEach(() => {
    vi.useFakeTimers({
      now: CURRENT_TIME,
    });
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('prints the logs to the writable stream', () => {
    const { stream, getOutput } = createStream();
    const backend = createJsonBackend({ stream });

    const logger = createLogger(backend);
    logger.debug('shout');
    logger.info('normal');
    logger.warn('hmmmm');
    logger.error('oh no');

    expect(getOutput()).toMatchSnapshot();
  });

  it('escapes newlines and carriage returns', () => {
    const { stream, getOutput } = createStream();
    const backend = createJsonBackend({ stream });
    const logger = createLogger(backend);

    logger.debug('sneaky log\nwith newlines\rand carriage returns\r\n');

    expect(getOutput()).toContain(
      'sneaky log\\nwith newlines\\rand carriage returns\\r\\n'
    );
  });

  it('includes log context', () => {
    const { stream, getOutput } = createStream();
    const backend = createJsonBackend({ stream });
    const logger = createLogger(backend);

    logger.debug('hello', { reqId: 'abc' });

    expect(getOutput()).toContain('"reqId":"abc"');
  });
});
