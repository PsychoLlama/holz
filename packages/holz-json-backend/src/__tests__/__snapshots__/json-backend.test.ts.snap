// Vitest Snapshot v1

exports[`JSON backend > prints the logs to the writable stream 1`] = `
"{\\"level\\":\\"debug\\",\\"time\\":\\"2020-06-15T12:00:00.000Z\\",\\"msg\\":\\"shout\\"}
{\\"level\\":\\"info\\",\\"time\\":\\"2020-06-15T12:00:00.000Z\\",\\"msg\\":\\"normal\\"}
{\\"level\\":\\"warn\\",\\"time\\":\\"2020-06-15T12:00:00.000Z\\",\\"msg\\":\\"hmmmm\\"}
{\\"level\\":\\"error\\",\\"time\\":\\"2020-06-15T12:00:00.000Z\\",\\"msg\\":\\"oh no\\"}
"
`;
